### PR TITLE
Integration of Hyperledger Burrow EVM

### PR DESCRIPTION
**Pull request by ekovalev**
### Description of the Change

This PR introduces changes necessary for integration of the Hyperledger Burrow EVM as a static library.

### Benefits

Ability to run smart contracts on top of Iroha blockchain.

### Possible Drawbacks 

Larger binaries.

### CODE REVIEW

**Comment by vadimreutskiy** [comment]
Burrow storage (in Burrow) — Eugene
Burrow storage (in Iroha) — Eugene
Query — Mikhail
Command — Eugene
Go part — Alexander
All other small parts — we will decide according to review from Andrey

**Comment by MBoldyrev** [comment]
Partial review progress:
**Go code**:
- [ ] Alexander

**Query**:
- [ ] Mikhail

**Comand**:
- [ ] Eugene

**Comment by MBoldyrev** [comment]
> Burrow storage (in Burrow) — Eugene

it is written mostly by Eugene and is hardly distinguishable from the whole go part IMHO

> Burrow storage (in Iroha) — Eugene

this is actually in #493

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp [hunk 1]
```diff
@@ -1375,6 +1403,183 @@ namespace iroha {
...
+                    shared_model::interface::types::EvmAddressHexString,
+                    shared_model::interface::types::EvmDataHexString,
+                    uint32_t,
+                    shared_model::interface::types::EvmAddressHexString,
+                    shared_model::interface::types::EvmDataHexString,
+                    shared_model::interface::types::EvmTopicsHexString>;
+
+      using PermissionTuple = boost::tuple<int>;
+
+      return executeQuery<QueryTuple, PermissionTuple>(
```
Comment (lebdron) [inline irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:1459]: "No transaction with such hash" error might be useful

File: irohad/ametsuchi/vm_caller.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  class session;
+}
+
+namespace iroha::ametsuchi {
+  class CommandExecutor;
+  class SpecificQueryExecutor;
+
+  class VmCaller {
+   public:
+    virtual iroha::expected::Result<std::string, std::string> call(
```
Comment (lebdron) [inline irohad/ametsuchi/vm_caller.hpp:28]: Documentation would be great

File: irohad/ametsuchi/vm_caller.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_AMETSUCHI_VM_CALLER_HPP
+#define IROHA_AMETSUCHI_VM_CALLER_HPP
+
+#include <functional>
+#include <memory>
```
Comment (lebdron) [inline irohad/ametsuchi/vm_caller.hpp:10]: These do not seem to be used

File: libs/common/hexutils.hpp [hunk 4]
```diff
@@ -75,6 +75,20 @@ namespace iroha {
...
+  /**
+   * Convert uint64_t number to a printable hex string
+   * @param val - unsinged integer value
+   * @return - converted hex string
+   */
+  inline std::string uint64ToHexstring(const uint64_t val) {
+    std::stringstream ss;
+    ss << std::hex << val;
+    auto res = ss.str();
+    if (res.size() & 0x1)
```
Comment (lebdron) [inline libs/common/hexutils.hpp:87]: Maybe use `std::setfill ('0') << std::setw(sizeof(uint64_t)*2) ` instead?

File: shared_model/backend/plain/engine_log.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+      EngineLog(interface::types::EvmAddressHexString const &address,
+                interface::types::EvmDataHexString const &data)
+          : address_(address), data_(data) {}
+
+      EngineLog(interface::types::EvmAddressHexString &&address,
+                interface::types::EvmDataHexString &&data)
+          : address_(std::move(address)), data_(std::move(data)) {}
+
+      interface::types::EvmAddressHexString const &getAddress() const {
```
Comment (lebdron) [inline shared_model/backend/plain/engine_log.hpp:30]: No cpp :(

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp [hunk 6]
```diff
@@ -1375,6 +1395,185 @@ namespace iroha {
...
+                prev_cmd_ix;
+
+            auto store_record = [](RecordsCollection &records,
+                                   RecordPtr &&rec) {
+              if (!!rec) {
+                records.emplace_back(std::move(rec));
+              }
+            };
+
+            auto store_log = [](RecordPtr &rec, EngineLogPtr &&el) {
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:1486]: I don't see the profit to use lambda context instead of arguments transmission.

File: shared_model/backend/plain/engine_log.hpp (hunk 5 above)
Comment (iceseer) [reply shared_model/backend/plain/engine_log.hpp:30]: Yep :)

File: goSrc/src/vmCaller/iroha/storage.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	}
+}
+
+func handleIrohaCallResult(result C.Iroha_Result) (*string, error) {
+	switch result.which {
+	case C.Iroha_Result_Type_Value:
+		return result.data.toStringAndRelease(), nil
+	case C.Iroha_Result_Type_Error:
+		err_str := result.data.toStringAndRelease()
+		if err_str != nil {
```
[the remark this answers is not in the document]
Comment (MBoldyrev) [reply goSrc/src/vmCaller/iroha/storage.go:34]: there is a "catch-all" return of error with `"unknown error"` message, does not it do what you mean?

File: goSrc/src/vmCaller/evm/native_contract.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+		}
+	}
+
+	ctx.Logger.Trace.Log("function", "getAssetBalance",
+		"account", args.Account,
+		"asset", args.Asset,
+		"value", value)
+
+	return getAssetBalanceRets{Result: value}, nil
+}
```
[the remark this answers is not in the document]
Comment (lebdron) [reply goSrc/src/vmCaller/evm/native_contract.go:72]: @ekovalev We decided to return "0" instead of ""

File: goSrc/src/vmCaller/iroha/storage.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+			return nil, errors.Errorf(errors.Codes.ExecutionReverted, *err_str)
+		}
+	}
+	return nil, errors.Errorf(errors.Codes.ExecutionReverted, "unknown error")
+}
+
+func (st *IrohaStorage) GetAccount(address crypto.Address) (*acm.Account, error) {
+	cAddress := MakeIrohaCharBuffer(address.String())
+	accountBytesHex, err := handleIrohaCallResult(C.Iroha_GetAccount(st.storage, *cAddress))
+	cAddress.free()
```
[the remark this answers is not in the document]
Comment (lebdron) [reply goSrc/src/vmCaller/iroha/storage.go]: Since both sides are represented by C interfaces, exceptions should be handled in C++ code and return as error code or something else to go through C interface. Same reasoning for go error handling.

File: goSrc/src/vmCaller/iroha/commands.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+			GetAccountAssets: &pb.GetAccountAssets{AccountId: accountID}}}}
+	queryResponse, err := makeProtobufQueryAndExecute(IrohaQueryExecutor, query)
+	if err != nil {
+		return []*pb.AccountAsset{}, err
+	}
+	switch response := queryResponse.Response.(type) {
+	case *pb.QueryResponse_ErrorResponse:
+		if response.ErrorResponse.Reason == pb.ErrorResponse_NO_ACCOUNT {
+			// No errors, but requested account does not exist
+			return []*pb.AccountAsset{}, nil
```
Comment (lebdron) [inline goSrc/src/vmCaller/iroha/commands.go:73]: Will this result in "0" returned in native getAssetBalance contract for non-existing account?

Comment (ekovalev) [reply goSrc/src/vmCaller/iroha/commands.go]: not anymore

File: test/framework/executor_itf/executor_itf.hpp [hunk 11]
```diff
@@ -66,7 +66,7 @@ namespace iroha {
       iroha::ametsuchi::CommandResult executeCommandAsAccount(
           const shared_model::interface::Command &cmd,
           const std::string &account_id,
-          bool do_validation) const;
```
Comment (lebdron) [inline test/framework/executor_itf/executor_itf.hpp:69]: Decided to revert these changes @MBoldyrev
